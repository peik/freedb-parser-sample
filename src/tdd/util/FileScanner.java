package tdd.util;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FilenameFilter;
import java.util.Arrays;
import java.util.List;

/**
 * FileScanner lists all files in a given directory.
 */
public class FileScanner {

	/** Main directory containing the files to list */
	private final String dir;

	public FileScanner(String dir) {
		this.dir = dir;
	}

	public List<String> listTxtFiles() throws FileNotFoundException {
		File dir = new File(this.dir);

		String[] txtFiles = dir.list(this.txtFilter);
		if (txtFiles == null) {
			throw new FileNotFoundException("Directory " + dir + " not found");
		}

		return Arrays.asList(txtFiles);
	}

	// Anonymous FilenameFilter class to find only .txt files
	private final FilenameFilter txtFilter = new FilenameFilter() {
		public boolean accept(File dir, String name) {
			return name.endsWith(".txt");
		}
	};

}
