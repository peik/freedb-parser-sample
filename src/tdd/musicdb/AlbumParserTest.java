package tdd.musicdb;

import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Test;

import com.thoughtworks.xstream.XStream;

public class AlbumParserTest {

	@Test
	public void testCalculateSongLengths() throws Exception {
		// Data copied from 12025d03.txt
		AlbumParser parser = new AlbumParser();
		parser.offsets.add(150);
		parser.offsets.add(12143);
		parser.offsets.add(31682);
		parser.discLengthSecs = 607;
		List<Integer> lengthsSecs = parser.calculateSongLengths();
		Approvals.approve(lengthsSecs.toString());
		// Expected result:
		// 1. I'm Your Puppet 2:39 = 159s
		// 2. Me And Jacky 4:20 = 260s
		// 3. If You Don't Like Country 3:05 = 185s
		// Total time: 10:04
	}

	@Test
	public void testParse() throws Exception {
		AlbumParser parser = new AlbumParser("c00ccd0d.txt");
		Approvals.approve(new XStream().toXML(parser));
	}

	@Test
	public void testParseToListing() throws Exception {
		AlbumParser parser = new AlbumParser("c00ccd0d.txt");
		Album album = parser.getAlbum();
		Approvals.approve(album.getListing());
	}

}
