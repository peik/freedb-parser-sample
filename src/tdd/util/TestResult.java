package tdd.util;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.approvaltests.Approvals;

import com.thoughtworks.xstream.XStream;
import com.thoughtworks.xstream.io.xml.XppDomDriver;

/**
 * A test helper class that collects the output of a test to be able to print it
 * in text form.<br>
 * Intended to collect data to be used with the {@link Approvals} tests library
 * (http://approvaltests.sourceforge.net/).
 * 
 * @author HiQ Finland Oy (9518174)
 * @version $Id: TestResult.java 8543 2011-12-16 16:23:58Z pas $
 */
public class TestResult implements java.io.Serializable {

	private static final long serialVersionUID = 1L;
	private final List<String> output = new ArrayList<String>();
	private boolean logToConsole = false;

	/**
	 * Constructor
	 */
	public TestResult() {
	}

	/**
	 * Constructor with header text.
	 */
	public TestResult(String header) {
		log(header);
	}

	/**
	 * By default, the result is only logged to the output list, from which it
	 * can be retrieved at the end of the test using {@link #getOutputText()}.
	 * For debugging purposes, you might want to enable logging also to the
	 * console, which prints out the output to System.out as soon as it is
	 * created, not just at the end of the test.
	 */
	public void enableLogToConsole(boolean logAlsoToConsole) {
		this.logToConsole = logAlsoToConsole;
	}

	/**
	 * Log a new message to the result.
	 */
	public void log(String message) {
		if (logToConsole) {
			System.out.println(message);
		}
		output.add(message);
	}

	/**
	 * Log a new message to the result, formatting the arguments using
	 * {@link String#format(String, Object...)}.
	 */
	public void log(String formatMessage, Object... args) {
		log(String.format(formatMessage, args));
	}

	/**
	 * Log a collection of objects using their <code>#toString()</code> values.
	 */
	public void logCollection(Collection<?> objects) {
		for (Object object : objects) {
			log(String.valueOf(object));
		}
	}

	/**
	 * Append a part of a message to the end of the previous logged message. No
	 * spaces or padding is added, so this is up to the responsibility of the
	 * caller.
	 * <p>
	 * This method is useful when you want to build the message conditionally,
	 * and append more infor to the message only under certain condition, for
	 * instance after checking that some object to write to the output is not
	 * null.
	 */
	public void append(String messagePart) {
		if (logToConsole) {
			System.out.println("... " + messagePart);
		}
		int lastPos = output.size() - 1;
		String lastLoggedMessage = output.get(lastPos);
		output.set(lastPos, lastLoggedMessage + messagePart);
	}

	/**
	 * Append a part of a message to the end of the previous logged message,
	 * formatting the arguments using {@link String#format(String, Object...)}.
	 * 
	 * @see #append(String)
	 */
	public void append(String formatMessagePart, Object... args) {
		append(String.format(formatMessagePart, args));
	}

	public void logXML(Object object) {
		XStream xs = new XStream(new XppDomDriver());
		String xml = xs.toXML(object);
		log(xml);
	}

	public void logXMLCollection(Collection<?> objects) {
		XStream xs = new XStream(new XppDomDriver());
		for (Object object : objects) {
			String xml = xs.toXML(object);
			log(xml);
		}
	}

	/**
	 * Add the output of the given result to this one, joining the results
	 * together.
	 */
	public void addOutput(TestResult result) {
		this.output.addAll(result.output);
	}

	public List<String> getOutput() {
		return output;
	}

	public String getOutputText() {
		StringBuilder sb = new StringBuilder();
		for (String line : output) {
			sb.append(line).append("\n");
		}
		return sb.toString();
	}

	/**
	 * Shortcut for <code>Approvals.approve(getOutputText());</code> which
	 * should be called at the end of an {@link Approvals} test.
	 */
	public void approve() throws Exception {
		Approvals.approve(getOutputText());
	}

}
