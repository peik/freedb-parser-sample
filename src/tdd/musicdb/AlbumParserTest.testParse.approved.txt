<tdd.musicdb.AlbumParser>
  <data class="tree-map">
    <no-comparator/>
    <entry>
      <string>DGENRE</string>
      <string>Pop</string>
    </entry>
    <entry>
      <string>DISCID</string>
      <string>c00ccd0d</string>
    </entry>
    <entry>
      <string>DTITLE</string>
      <string>Adele / 21</string>
    </entry>
    <entry>
      <string>DYEAR</string>
      <string>2011</string>
    </entry>
    <entry>
      <string>EXTD</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT0</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT1</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT10</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT11</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT12</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT2</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT3</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT4</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT5</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT6</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT7</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT8</string>
      <string></string>
    </entry>
    <entry>
      <string>EXTT9</string>
      <string></string>
    </entry>
    <entry>
      <string>PLAYORDER</string>
      <string></string>
    </entry>
    <entry>
      <string>TTITLE0</string>
      <string>Rolling In The Deep</string>
    </entry>
    <entry>
      <string>TTITLE1</string>
      <string>Rumour Has It</string>
    </entry>
    <entry>
      <string>TTITLE10</string>
      <string>Someone Like You</string>
    </entry>
    <entry>
      <string>TTITLE11</string>
      <string>If It Hadn&apos;t Been For Love</string>
    </entry>
    <entry>
      <string>TTITLE12</string>
      <string>Hiding My Heart</string>
    </entry>
    <entry>
      <string>TTITLE2</string>
      <string>Turning Tables</string>
    </entry>
    <entry>
      <string>TTITLE3</string>
      <string>Don&apos;t You Remember</string>
    </entry>
    <entry>
      <string>TTITLE4</string>
      <string>Set Fire To The Rain</string>
    </entry>
    <entry>
      <string>TTITLE5</string>
      <string>He Won&apos;t Go</string>
    </entry>
    <entry>
      <string>TTITLE6</string>
      <string>Take It All</string>
    </entry>
    <entry>
      <string>TTITLE7</string>
      <string>I&apos;ll Be Waiting</string>
    </entry>
    <entry>
      <string>TTITLE8</string>
      <string>One And Only</string>
    </entry>
    <entry>
      <string>TTITLE9</string>
      <string>Lovesong</string>
    </entry>
  </data>
  <offsets>
    <int>150</int>
    <int>17347</int>
    <int>34092</int>
    <int>52852</int>
    <int>71092</int>
    <int>89219</int>
    <int>110065</int>
    <int>127181</int>
    <int>145305</int>
    <int>171417</int>
    <int>195138</int>
    <int>216406</int>
    <int>230516</int>
  </offsets>
  <discLengthSecs>3279</discLengthSecs>
</tdd.musicdb.AlbumParser>